In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../Data/Raw/Elimited_sales_data.csv")

In [3]:
df

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,567297,84406B,CREAM CUPID HEARTS COAT HANGER,9,11/15/2025 19:30,2.75,15911,United Kingdom
1,567163,22197,POPCORN HOLDER,10,11/16/2025 7:24,0.85,13228,Netherlands
2,554498,23084,RABBIT NIGHT LIGHT,26,11/22/2025 15:12,2.08,12745,Eire
3,551304,20725,LUNCH BAG RED RETROSPOT,41,8/2/2025 23:33,1.65,12954,Portugal
4,537146,71053,WHITE METAL LANTERN,35,12/29/2025 10:54,3.39,18237,Germany
...,...,...,...,...,...,...,...,...
495,555058,POST,POSTAGE,6,9/14/2025 18:53,18.00,13523,United Kingdom
496,550503,84879,ASSORTED COLOUR BIRD ORNAMENT,24,11/4/2025 5:11,1.69,13437,United Kingdom
497,538534,POST,POSTAGE,19,9/20/2025 19:46,18.00,17325,United Kingdom
498,554921,47566,PARTY BUNTING,16,9/7/2025 13:41,4.65,18120,United Kingdom


## data-quality checks

In [4]:
df.shape

(500, 8)

In [5]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

In [6]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   InvoiceNo    500 non-null    int64  
 1   StockCode    500 non-null    str    
 2   Description  500 non-null    str    
 3   Quantity     500 non-null    int64  
 4   InvoiceDate  500 non-null    str    
 5   UnitPrice    500 non-null    float64
 6   CustomerID   500 non-null    int64  
 7   Country      500 non-null    str    
dtypes: float64(1), int64(3), str(4)
memory usage: 58.0 KB


In [9]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [10]:
df["InvoiceDate"].dtype

dtype('<M8[us]')

In [11]:
df[["Quantity", "UnitPrice"]].describe()

,Quantity,UnitPrice
count,500.000000,500.000000
mean,38.706000,4.865880
std,65.912096,5.268293
min,1.000000,0.550000
25%,15.000000,1.690000
50%,27.000000,2.080000
75%,40.000000,4.650000
max,493.000000,18.000000


## creating Revenue column

In [12]:
df["Revenue"] = df["Quantity"] * df["UnitPrice"]

In [13]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,567297,84406B,CREAM CUPID HEARTS COAT HANGER,9,2025-11-15 19:30:00,2.75,15911,United Kingdom,24.75
1,567163,22197,POPCORN HOLDER,10,2025-11-16 07:24:00,0.85,13228,Netherlands,8.50
2,554498,23084,RABBIT NIGHT LIGHT,26,2025-11-22 15:12:00,2.08,12745,Eire,54.08
3,551304,20725,LUNCH BAG RED RETROSPOT,41,2025-08-02 23:33:00,1.65,12954,Portugal,67.65
4,537146,71053,WHITE METAL LANTERN,35,2025-12-29 10:54:00,3.39,18237,Germany,118.65


## filtering with multiple conditions

In [14]:
df[(df["Quantity"] > 20) & (df["UnitPrice"] > 3)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
4,537146,71053,WHITE METAL LANTERN,35,2025-12-29 10:54:00,3.39,18237,Germany,118.65
6,543188,71053,WHITE METAL LANTERN,26,2025-07-21 01:44:00,3.39,18017,United Kingdom,88.14
8,576326,71053,WHITE METAL LANTERN,38,2025-08-11 07:51:00,3.39,13728,United Kingdom,128.82
13,549930,POST,POSTAGE,34,2025-02-09 08:35:00,18.00,12765,United Kingdom,612.00
14,556719,71053,WHITE METAL LANTERN,27,2025-11-18 22:27:00,3.39,16565,United Kingdom,91.53
...,...,...,...,...,...,...,...,...,...
469,545132,47566,PARTY BUNTING,38,2025-11-04 16:26:00,4.65,15003,Portugal,176.70
485,564591,22752,SET 7 BABUSHKA NESTING BOXES,32,2025-03-01 00:00:00,8.50,17012,United Kingdom,272.00
488,578503,47566,PARTY BUNTING,40,2025-06-28 01:34:00,4.65,14636,United Kingdom,186.00
490,573591,POST,POSTAGE,40,2025-05-13 03:47:00,18.00,14954,United Kingdom,720.00


## groupby()

In [15]:
country_revenue = df.groupby("Country")["Revenue"].sum()
country_revenue

Country
Australia           629.74
Belgium             487.30
Eire               1045.92
France             2291.31
Germany            2784.23
Netherlands        1380.17
Portugal           1475.93
Spain               329.02
Switzerland        1273.55
United Kingdom    52766.81
Name: Revenue, dtype: float64

## mean()

In [16]:
country_avg_revenue = df.groupby("Country")["Revenue"].mean()
country_avg_revenue

Country
Australia          52.478333
Belgium            81.216667
Eire               69.728000
France            134.782941
Germany            99.436786
Netherlands       172.521250
Portugal          134.175455
Spain              54.836667
Switzerland       141.505556
United Kingdom    135.996933
Name: Revenue, dtype: float64

## count()

In [17]:
country_transaction_count = df.groupby("Country")["InvoiceNo"].count()
country_transaction_count

Country
Australia          12
Belgium             6
Eire               15
France             17
Germany            28
Netherlands         8
Portugal           11
Spain               6
Switzerland         9
United Kingdom    388
Name: InvoiceNo, dtype: int64

## min()

In [18]:
country_min_revenue = df.groupby("Country")["Revenue"].min()
country_min_revenue

Country
Australia          2.75
Belgium           34.85
Eire               4.10
France             8.32
Germany            1.70
Netherlands        8.50
Portugal          34.65
Spain             21.97
Switzerland       10.25
United Kingdom     1.65
Name: Revenue, dtype: float64

## 
max()

In [19]:
country_max_revenue = df.groupby("Country")["Revenue"].max()
country_max_revenue

Country
Australia         144.00
Belgium           178.50
Eire              318.75
France            522.75
Germany           432.00
Netherlands       702.00
Portugal          576.00
Spain              94.30
Switzerland       522.00
United Kingdom    864.00
Name: Revenue, dtype: float64

## agg()

In [21]:
country_revenue_summary = df.groupby("Country")["Revenue"].agg(
    ["sum", "mean", "min", "max"]
)

country_revenue_summary

,sum,mean,min,max
Country,,,,
Australia,629.74,52.478333,2.75,144.00
Belgium,487.30,81.216667,34.85,178.50
Eire,1045.92,69.728000,4.10,318.75
France,2291.31,134.782941,8.32,522.75
Germany,2784.23,99.436786,1.70,432.00
Netherlands,1380.17,172.521250,8.50,702.00
Portugal,1475.93,134.175455,34.65,576.00
Spain,329.02,54.836667,21.97,94.30
Switzerland,1273.55,141.505556,10.25,522.00


## sort_values()

In [22]:
country_revenue.sort_values(ascending=False)

Country
United Kingdom    52766.81
Germany            2784.23
France             2291.31
Portugal           1475.93
Netherlands        1380.17
Switzerland        1273.55
Eire               1045.92
Australia           629.74
Belgium             487.30
Spain               329.02
Name: Revenue, dtype: float64

## value_counts()

In [23]:
df["Country"].value_counts()

Country
United Kingdom    388
Germany            28
France             17
Eire               15
Australia          12
Portugal           11
Switzerland         9
Netherlands         8
Belgium             6
Spain               6
Name: count, dtype: int64

## Working with dates

In [24]:
df["Year"] = df["InvoiceDate"].dt.year
df["Month"] = df["InvoiceDate"].dt.month
df["Day"] = df["InvoiceDate"].dt.day

In [25]:
df[["InvoiceDate", "Year", "Month", "Day"]].head()

,InvoiceDate,Year,Month,Day
0,2025-11-15 19:30:00,2025,11,15
1,2025-11-16 07:24:00,2025,11,16
2,2025-11-22 15:12:00,2025,11,22
3,2025-08-02 23:33:00,2025,8,2
4,2025-12-29 10:54:00,2025,12,29


## Practical 1

Question 1: What is the total revenue?

In [26]:
total_revenue = df["Revenue"].sum()
total_revenue

np.float64(64463.979999999996)

Interpretation: The total revenue generated from all 500 transactions was 64,463.98. This represents the combined revenue from all products and countries in the dataset.

Question 2: What is the average transaction value?

In [27]:
average_transaction_value = df["Revenue"].mean()
average_transaction_value

np.float64(128.92795999999998)

Interpretation: The average transaction value was 128.93. This means that, on average, each transaction generated approximately 128.93 in revenue.

Question 3: Which 10 products generate the most revenue?

In [29]:
top_10_products = (
    df.groupby("Description")["Revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_10_products

Description
POSTAGE                               18414.00
REGENCY CAKESTAND 3 TIER              11959.50
SET 7 BABUSHKA NESTING BOXES           8126.00
PARTY BUNTING                          3613.05
POPCORN HOLDER                         3031.10
WHITE HANGING HEART T-LIGHT HOLDER     2394.45
PACK OF 72 RETROSPOT CAKE CASES        2394.15
WHITE METAL LANTERN                    2356.05
RABBIT NIGHT LIGHT                     2325.44
CREAM CUPID HEARTS COAT HANGER         2271.50
Name: Revenue, dtype: float64

Interpretation: POSTAGE generated the highest revenue at 18,414.00, followed by REGENCY CAKESTAND 3 TIER at 11,959.50 and SET 7 BABUSHKA NESTING BOXES at 8,126.00. The results show that revenue was concentrated among a small number of products, with POSTAGE contributing the highest amount.

Question 4: Which 5 countries generate the most revenue?

In [30]:
top_5_countries = (
    df.groupby("Country")["Revenue"]
    .sum()
    .sort_values(ascending=False)
    .head(5)
)

top_5_countries

Country
United Kingdom    52766.81
Germany            2784.23
France             2291.31
Portugal           1475.93
Netherlands        1380.17
Name: Revenue, dtype: float64

Interpretation: The United Kingdom generated the highest revenue at 52,766.81, far exceeding the other four countries. Germany ranked second with 2,784.23, while the Netherlands recorded the lowest revenue among the top five at 1,380.17.

Question 5: Which products sell the highest quantities?

In [31]:
top_10_quantity_products = (
    df.groupby("Description")["Quantity"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_10_quantity_products

Description
PACK OF 72 RETROSPOT CAKE CASES       4353
POPCORN HOLDER                        3566
RABBIT NIGHT LIGHT                    1118
POSTAGE                               1023
ASSORTED COLOUR BIRD ORNAMENT          986
SET 7 BABUSHKA NESTING BOXES           956
WHITE HANGING HEART T-LIGHT HOLDER     939
REGENCY CAKESTAND 3 TIER               938
JUMBO BAG VINTAGE DOILY                918
LUNCH BAG SUKI DESIGN                  843
Name: Quantity, dtype: int64

Interpretation: PACK OF 72 RETROSPOT CAKE CASES had the highest quantity sold, with 4,353 units, followed by POPCORN HOLDER with 3,566 units. These products had the highest sales volume in terms of quantity within the dataset.

Question 6: Which month generated the highest revenue?

In [32]:
monthly_revenue = (
    df.groupby("Month")["Revenue"]
    .sum()
    .sort_values(ascending=False)
)

monthly_revenue

Month
6     7237.98
3     6607.64
12    6346.28
10    6072.97
5     5530.88
8     5408.85
4     5345.51
1     5034.99
11    4744.68
7     4572.21
9     4152.20
2     3409.79
Name: Revenue, dtype: float64

Interpretation: June generated the highest revenue at 7,237.98, while February generated the lowest revenue at 3,409.79. Revenue varied across the months, with June showing the strongest monthly performance.